Convert BME280 to multi-instance, use new DT API

mbolivar-nordic opened the pull request:
Make some sample changes as well.

Tested with two devices at the same time, one on each bus.

mbolivar-nordic committed:
samples: bme280: enable sensor logging

This helps debug issues with the device. Samples should try to be
helpful to first-time users. Send printk() to logging so it doesn't
fight over the UART.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: bme280: nrf52840 DK: set LSB in i2c addr

The nrf52840-DK doesn't have this sensor built-in, so it's likely
users will be getting this sensor from a breakout board. The ones that
Sparkfun and Adafruit sell pull up the LSB address pin, so the default
address is 0x77 on those boards.

Let's follow along with them instead of using 0x76, because those are
the boards that come up first in the results when you google "BME280
breakout board" from where I'm sitting.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
```
$ wc -l /home/mbolivar/zp/zephyr/sanity-out/frdm_k64f/tests/drivers/build_all/sensors.build.a_h/build.log
78041 /home/mbolivar/zp/zephyr/sanity-out/frdm_k64f/tests/drivers/build_all/sensors.build.a_h/build.log
```

That's an impressive error message. We're going to have to do something about making these shorter.

gmarull commented on drivers/sensor/bme280/bme280.c:
```diff
@@ -306,95 +370,187 @@ static int bme280_chip_init(struct device *dev)
+		.bus_config =  { .i2c_addr = DT_INST_REG_ADDR(inst), }	\
+	}
+
+/* TODO replace with a foreach over all enabled devices when available */
```
@mbolivar-nordic when will the `foreach` macro be available?

mbolivar-nordic replied:
There is a PR. When it's merged it'll be ready :).

gmarull replied:
seen #23416 thanks!

mbolivar-nordic commented:
Rebased on top of https://github.com/zephyrproject-rtos/zephyr/pull/23792, DNM until that has a chance to be reviewed

mbolivar-nordic commented:
> Rebased on top of #23792, DNM until that has a chance to be reviewed

Rebased on master, dropping DNM

mbolivar-nordic committed:
drivers: sensor: bme280: convert to new DT API

Make this multi-instance and use the new API.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: bme280: update to new DT API

Use instance zero of the compatible to get the device instead of
fixing a label. Print the used label for help debugging, and the
results of finding the device (or not), with a hint about what might
have gone wrong in case of failure.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
Rebased again

galak commented on drivers/sensor/bme280/bme280.h:
```diff
@@ -10,6 +10,11 @@
 #include <device.h>
+#include <devicetree.h>
+
+#define DT_DRV_COMPAT bosch_bme280
```
I was avoid putting DT_DRV_COMPAT in header if possible.

galak commented on drivers/sensor/bme280/bme280.c:
```diff
@@ -305,113 +370,213 @@ static int bme280_read_compensation(struct bme280_data *data)
+/* This is only usable if BME280_ON_SPI(inst) */
+#define BME280_HAS_CS(inst) DT_INST_SPI_DEV_HAS_CS_GPIOS(inst)
+
+#define BME280_DEFINE(inst)						\
```
Can you move this below `BME280_DEFINE_SPI` and `BME280_DEFINE_I2C`

galak reviewed (approved):
In general looks fine.  Would like to see the macro's related to BME280_DEFINE re-organized and have more comments as I imagine we'll want to point people at this as an example of how to do things.

mbolivar-nordic committed:
scripts: kconfig: add dt_compat_on_bus() helper

This can be used from Kconfig to detect if any enabled nodes of a
compatible are on a bus. This can be useful if a compatible might
appear on multiple buses, to enable them by default from application
code without having to change prj.conf settings depending on what's in
the devicetree.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: bme280: enable the appropriate bus by default

I personally don't find it very useful to have to maintain prj.conf
and prj_spi.conf for this sample. The information we need to make
this application "just work" with regards to the bus is available in
the devicetree, and Kconfig can now access it using
dt_compat_on_bus().

Do so, enabling I2C and SPI appropriately when a sensor of the right
type is on either of those buses.

If no sensors are enabled, the user gets the build-time error message
in main.c about no devices being found.

This approach is prone to the "stuck symbol" Kconfig problem covered
in our documentation, so a pristine build is necessary to change the
default settings from a previous build.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@galak review addressed, thank you. I've also added a new dt_compat_on_bus() helper. I kept getting annoyed at having to flip between prj.conf and prj_spi.conf, so the same just default-enables the right bus now using that.
